Objects/setobject.c

```diff
@@ -2156,13 +2323,17 @@ set_init(PySetObject *self, PyObject *args, PyObject *kwds)
 {
     PyObject *iterable = NULL;
 
-     if (!_PyArg_NoKeywords("set", kwds))
+    if (!_PyArg_NoKeywords("set", kwds))
         return -1;
     if (!PyArg_UnpackTuple(args, Py_TYPE(self)->tp_name, 0, 1, &iterable))
         return -1;
+
+    Py_BEGIN_CRITICAL_SECTION(self);
```

Comment:
Maybe it's worth inlining `set_update_internal` into here?  It'd mean we'd only take the critical section once, and also so that the clear and update are done atomically.  As of now someone could call `__init__` on multiple threads and end up with elements from both callers.